lib/Test/Mockify/Injector.pm

```diff
@@ -0,0 +1,128 @@
+package Test::Mockify::Injector;
```

Comment:
This alternative doesn't provide much value that I can see beyond the current solution, or maybe I completely missed your point. MockScope seems to simply aggregate the various mocked packages into one place, though I don't see that being terribly useful or valuable. MockScope does provide explicit calls (clearPackage, clearAll) to restore the original functionality; however, if that is truly desired, we could easily add that to the existing implementation. Compare your implementation to what it would currently be:

```perl
my $packageNameInjector = Test::Mockify::Injector->new('Package::Name');
$packageNameInjector->mock('MethodName')->whenAny()->thenReturnUndef();
my $packageTwoInjector = Test::Mockify::Injector->new('Package::Two');
$packageTwoInjector->mock('one')->when(String('hello'))->thenReturn('world');
$packageTwoInjector->spy('two')->whenAny();

# actual test    #############################################################
my $Sut = Sut->new();
$Sut->callWithStaticDependenies();
##############################################################################
```

There is not a lot of difference. If you want the original implementation restored at the end, simply wrap that code in a block `{...}`, or we could add a `restore` function to the Injector so you can do that explicitly. I suppose if you're doing a bunch of mocking having the `clearAll` could be useful, but I doubt that will happen much.

```perl
$packageNameInjector->restore();
$packageTwoInjector->restore();
```

The current Injector implementation has just as much chaining capability as Mockify does. You could absolutely chain `Test::Mockify::Injector->new('Package::Name')->mock('MethodName')->whenAny()->thenReturn('ReturnValue')` if you don't care to have a reference to the Injector later (to mock/spy additional methods or for verification purposes). The same is true of Mockify.

Why don't you like Injector as a name? In the context of the way Mockify works and what the Injector is trying to change about that, I feel it adequately describes the functionality.

Reply:
In this case, the naming difficulty is less of a design issue in my opinion, and more the fact that it is difficult to capture in a succinct name the essence of what this module is trying to do.

Reply:
Yes, you are right. I didn't saw this possibility of chaining and this is absolutely enough. 
```perl
Test::Mockify::Injector->new('Package::Name')->mock('MethodName')->whenAny()->thenReturn('ReturnValue')
```

Reply:
So I would mainly keep what you did (if the base thing is solved), but the name is not acceptable for me. "Injection" is a verb. "Injector" is a nominalized verb of it.
But let me share my thoughts:
I think we agree on that the Injector gives the possibility to inject( and here is the verb correct) a mock into something.  The Question is what is something, so where we inject the mock. Inject something into the Injector make no sense for my point of view. I think we **inject** a **mock of a function** into a **Package** which is valid in our **current scope**. So i would suggest "Package" instead of "Injector". And even one step further we could rename the action(verb) from "mock" to "mockFunction"
```perl
{ # this defines the scope
    my PackageTool = Test::Mockify::Package->new('Some::Tool');
    PackageTooll->mockFunction('Greet')->when(String('Dustin'))->thenReturn('Hello Dustin');
    PackageTool->mockFunction('Greet')->when(String('Christian'))->thenReturn('Hello Christian');
    PackageTool->mockFunction('SayBye')->whenAny()->thenReturn('Bye All');
}
```  
what do you think?